File: src/main/java/am/adrian/dungeonkeeper/game/MoveValidator.java
package am.adrian.dungeonkeeper.game;

import am.adrian.dungeonkeeper.constant.Direction;
import am.adrian.dungeonkeeper.model.coords.Coords;
import am.adrian.dungeonkeeper.model.gameobject.Creature;
import am.adrian.dungeonkeeper.model.gameobject.GameObject;
import am.adrian.dungeonkeeper.model.gameobject.Path;
import am.adrian.dungeonkeeper.model.gameobject.Wall;
import am.adrian.dungeonkeeper.model.gameobject.tile.Water;
import am.adrian.dungeonkeeper.model.gameobject.trait.CanFly;
import am.adrian.dungeonkeeper.model.gameobject.trait.CanSwim;
import am.adrian.dungeonkeeper.model.gameobject.trait.CanWalk;
import org.springframework.stereotype.Component;

@Component
public record MoveValidator(GameMap map) {

    public boolean validateWalk(Creature character, Direction dir) {
        if (!(character instanceof CanWalk)) {
            return false;
        }
        return isMoveWithinMap(character, dir) && objectToMoveTo(character.getCoords(), dir) instanceof Path;
    }

    public boolean validateSwim(Creature character, Direction dir) {
        if (!(character instanceof CanSwim)) {
            return false;
        }
        return isMoveWithinMap(character, dir) && objectToMoveTo(character.getCoords(), dir) instanceof Water;
    }

    public boolean validateFly(Creature character, Direction dir) {
        if (!(character instanceof CanFly)) {
            return false;
        }
        return isMoveWithinMap(character, dir) && !(objectToMoveTo(character.getCoords(), dir) instanceof Wall);
    }

    private boolean isMoveWithinMap(Creature character, Direction dir) {
        return switch (dir) {
            case UP -> character.getCoords().y() > 0;
            case DOWN -> character.getCoords().y() < map.getHeight() - 1;
            case LEFT -> character.getCoords().x() > 0;
            case RIGHT -> character.getCoords().x() < map.getWidth() - 1;
        };
    }

    private GameObject objectToMoveTo(Coords characterCoords, Direction dir) {
        int x = characterCoords.x();
        int y = characterCoords.y();
        return switch (dir) {
            case UP -> map.getObjectMap()[y - 1][x];
            case DOWN -> map.getObjectMap()[y + 1][x];
            case LEFT -> map.getObjectMap()[y][x - 1];
            case RIGHT -> map.getObjectMap()[y][x + 1];
        };
    }
}

File: src/main/kotlin/am/adrian/dungeonkeeper/game/MoveValidator.kt
package am.adrian.dungeonkeeper.game

import am.adrian.dungeonkeeper.constant.Direction
import am.adrian.dungeonkeeper.model.coords.Coords
import am.adrian.dungeonkeeper.model.gameobject.GameObject
import am.adrian.dungeonkeeper.model.gameobject.creature.Creature
import am.adrian.dungeonkeeper.model.gameobject.tile.path.Path
import am.adrian.dungeonkeeper.model.gameobject.tile.path.Water
import am.adrian.dungeonkeeper.model.gameobject.tile.wall.Wall
import am.adrian.dungeonkeeper.model.gameobject.trait.CanFly
import am.adrian.dungeonkeeper.model.gameobject.trait.CanSwim
import am.adrian.dungeonkeeper.model.gameobject.trait.CanWalk
import org.springframework.stereotype.Component

@Component
class MoveValidator(private val map: GameMap) {

    fun validateWalk(character: Creature, dir: Direction): Boolean =
        if (character !is CanWalk) {
            false
        } else {
            isMoveWithinMap(character, dir) && objectToMoveTo(character.getCoords(), dir) is Path
        }

    fun validateSwim(character: Creature, dir: Direction): Boolean =
        if (character !is CanSwim) {
            false
        } else {
            isMoveWithinMap(character, dir) && objectToMoveTo(character.getCoords(), dir) is Water
        }

    fun validateFly(character: Creature, dir: Direction): Boolean =
        if (character !is CanFly) {
            false
        } else {
            isMoveWithinMap(character, dir) && objectToMoveTo(character.getCoords(), dir) !is Wall
        }

    private fun isMoveWithinMap(character: Creature, dir: Direction): Boolean {
        return when (dir) {
            Direction.UP -> character.getCoords().y() > 0
            Direction.DOWN -> character.getCoords().y() < map.height - 1
            Direction.LEFT -> character.getCoords().x() > 0
            Direction.RIGHT -> character.getCoords().x() < map.width - 1
        }
    }

    private fun objectToMoveTo(characterCoords: Coords, dir: Direction): GameObject? {
        val x = characterCoords.x()
        val y = characterCoords.y()
        return when (dir) {
            Direction.UP -> map.objectMap[y - 1][x]
            Direction.DOWN -> map.objectMap[y + 1][x]
            Direction.LEFT -> map.objectMap[y][x - 1]
            Direction.RIGHT -> map.objectMap[y][x + 1]
        }
    }
}
